### PR TITLE
Implement auto-inference for Java package repositories.

### LINKED ISSUE #20835: codeintel: implement an autoindexing recognizer for Java packages
The recognizer should: recognize LSIF packageInformation vertices which represent Maven repositories, request the repositories to be pulled down into gitserver, and submit an autoindex job to the queue which will index those repos if they haven't already been indexed.

### PR DESCRIPTION
**Pull request by olafurpg**
Fixes #20835. Previously, we had to manually add auto-indexing
configuration for Java package repositories. This commit implements
auto-inference for repos that have a `lsif-java.json` file at the root
directory. These repositories are created by the new JVM Dependencies
external service type and they should have a high likelyhood of indexing
successfully.

If this change works out well in production, we can consider expanding
the pattern list to also match Gradle/Maven builds.

### CODE REVIEW

**Comment by olafurpg** [comment]
Whoops, realized I forgot to update the tests to the new structure. Working on that right now, will undraft once the tests are done

File: lib/codeintel/autoindex/inference/java.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+			return true
+		}
+	}
+	return false
+}
+
+var supportedFilenames = []string{
+	// The "lsif-java.json" file is used to index package repositories such as
+	// the JDK sources and published Java libraries.
+	"lsif-java.json",
```
Comment (efritz) [inline lib/codeintel/autoindex/inference/java.go:56]: Assuming this is a file that the ext service puts there as part of the "clone" operation?

Comment (olafurpg) [reply lib/codeintel/autoindex/inference/java.go:56]: Correct, here's an example https://sourcegraph.com/maven/com.spotify/dns/-/blob/lsif-java.json

File: lib/codeintel/autoindex/inference/java_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		{paths: []string{"nested/build.gradle"}, expected: false},
+		{paths: []string{"settings.gradle"}, expected: true},
+		{paths: []string{"nested/settings.gradle"}, expected: false},
+		{paths: []string{"lsif-java.json"}, expected: true},
+		{paths: []string{"nested/lsif-java.json"}, expected: false},
+		{paths: []string{"build.sbt"}, expected: false},
+		{paths: []string{"package.json"}, expected: false},
+		{paths: []string{"MyApp.java"}, expected: false},
+		{paths: []string{"MyApp.groovy"}, expected: false},
+		{paths: []string{"gradle.properties"}, expected: false},
```
Comment (efritz) [inline lib/codeintel/autoindex/inference/java_test.go:31]: I'm surprised some of these are true (apart from lsif-java.json; do tests pass)?

File: lib/codeintel/autoindex/inference/java_test.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+		t.Run(name, func(t *testing.T) {
+			if value := recognizer.CanIndex(testCase.paths, NewMockGitserverClientWrapper()); value != testCase.expected {
+				t.Errorf("unexpected result from CanIndex. want=%v have=%v", testCase.expected, value)
+			}
+		})
+	}
+}
+
+func TestLsifJavaJobRecognizerInferIndexJobsTsConfigRoot(t *testing.T) {
```
Comment (efritz) [inline lib/codeintel/autoindex/inference/java_test.go:45]: Strange test name ... copy/paste?

**Comment by efritz** [comment]
I'm on board with this strategy, seems like a smart way to test without a huge blast radius.

**Comment by olafurpg** [comment]
OK, the tests have been fixed now.

**Review by efritz — APPROVED** [review]

**Comment by olafurpg** [comment]
Let's GOOOOOOOOOOOOOOOOOO 💪

**Pull request merged**